Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.Firestore.DistributedCache/FirestoreCache.cs

```diff
@@ -0,0 +1,323 @@
+﻿// Copyright (c) 2019 Google LLC.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+// 
+// http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+using System;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Google.Apis.Auth.OAuth2;
+using Google.Cloud.Firestore;
+using Microsoft.Extensions.Caching.Distributed;
+using Microsoft.Extensions.Logging;
+
+namespace Google.Cloud.AspNetCore.Firestore.DistributedCache
+{
+    /// <summary>
+    /// A distributed cache that stores entries in Firestore.
+    /// </summary>
+    public class FirestoreCache : IDistributedCache
+    {
+        private readonly FirestoreDb _firestore;
+        private readonly CollectionReference _cacheEntries;
+        private readonly ILogger<FirestoreCache> _logger;
+
+        public FirestoreCache(FirestoreDb firestore, ILogger<FirestoreCache> logger,
+            string collection = "CacheEntries")
+        {
+            if (logger is null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+
+            if (string.IsNullOrWhiteSpace(collection))
+            {
+                throw new ArgumentException("Must not be empty",
+                    nameof(collection));
+            }
+
+            _firestore = firestore;
+            _cacheEntries = _firestore.Collection(collection);
+            _logger = logger;
+        }
+        public FirestoreCache(string projectId, ILogger<FirestoreCache> logger,
+            string collection = "CacheEntries")
+            : this(FirestoreDb.Create(projectId), logger, collection)
+        {
+        }
+
+        public FirestoreCache(ILogger<FirestoreCache> logger,
+            string collection = "CacheEntries")
+            : this(FirestoreDb.Create(GetProjectId()), logger, collection)
+        {            
+        }
+
+        byte[] IDistributedCache.Get(string key) =>

```

Comment:
We should call `Task.Run` instead of just blocking on the task, otherwise it could deadlock. (Unlikely in ASP.NET Core, but...)

Reply:
I Googled `task wait deadlock` and read 3-4 articles like [this one](https://social.msdn.microsoft.com/Forums/vstudio/en-US/65a5f291-e4bf-4327-a4fa-b42b21add0e2/how-there-can-be-deadlock-in-taskwait?forum=csharpgeneral) and [that one](https://medium.com/rubrikkgroup/understanding-async-avoiding-deadlocks-e41f8f2c6f5d).

I see the potential for deadlock in `Get()` as described in the second article:

> The problem is that if you blocked the thread which is supposed to complete the I/O, then there won’t be a thread to complete the I/O.
> If called from threadpool thread then a theadpool thread is blocked, which will lead to a deadlock if the work load is high enough. If all threads are blocked in the threadpool then there will be nobody to complete the Task.

I don't understand is how calling `Task.Run(() =>_cacheEntries.Document(key).GetSnapshotAsync()).Result` will mitigate that risk.

Either way the calling thread must block, and and the GetSnapshot() code will schedule work on the threadpool.

Reply:
The problem is if a continuation tries to be scheduled in the same synchronization context which is blocked. It's mostly a problem for WinForms or WPF rather than ASP.NET Core, but I'd suggest it as a matter of course. (It's not ideal to use async code from sync, but often we just don't have a choice, like this...)